### DISCUSSION TITLE
Simple transformation

### DISCUSSION THREAD

**Discussion by ztef** [vsg-dev/VulkanSceneGraph#873] [General]
How to position geometries (a Cylinder) in a map using TileDatabase as the scene ?

I am experimenting using Builder and transforms, but i cannot draw a cylinder parallel to the surface as seen in the following 

![Screenshot from 2023-07-12 13-18-22](https://github.com/vsg-dev/VulkanSceneGraph/assets/1700294/7b8d7481-f440-4fd6-908d-0ec133eec883)

**Comment by ztef** [comment]
I am trying to rotate the cylinder 45 degrees to be perpendicular to the surface of the map

auto  rot = vsg::rotate(vsg::radians(45.0f), 1.0f, 0.0f, 0.0f);
geomInfo.transform = rot;
auto cyl = builder->createCylinder(geomInfo, stateInfo);

but the result is that the cylinder is drawn in another part of the world. Seems that takes the 45 degrees relative to the center of the world.

any suggestions ?

**Comment by robertosfield** [comment]
The vsg::EllipsoidModel is what provides the definition of the earths shape, you can use this to convert Lat, Long, Alititiude to Earth Centered Earth Fixed, you can also get transforms for local origins to ECEF and visa-versa.

**Comment by ztef** [comment]
Yes, i am using it this way :

 ellipsoidModel->convertLatLongAltitudeToECEF({25.686613, -100.316116, 0.0}); so i am able to position the cylinder using the lat long coordinates.  My Issue is the transformation to rotate the cylinder around its own axis to be oriented paralell to the surface.


When I rotate the cylnder, it rotates using the center of the earth as origin, but i need to rotate around its own.

**Comment by ztef** [comment]
This is the code : 

                            vsg::ref_ptr<vsg::TileDatabaseSettings> settings;
                        
                            settings = vsg::TileDatabaseSettings::create();
                            settings->extents = {{-180.0, -90.0, 0.0}, {180.0, 90.0, 1.0}};
                            settings->noX = 1;
                            settings->noY = 1;
                            settings->maxLevel = 17;
                            settings->originTopLeft = true;
                            settings->lighting = false;
                            settings->projection = "EPSG:3857"; // Spherical Mecator
                            settings->imageLayer = "http://a.tile.openstreetmap.org/{z}/{x}/{y}.png";

                            arguments.read("-t", settings->lodTransitionScreenHeightRatio);
                            arguments.read("-m", settings->maxLevel);

                            auto ellipsoidModel = settings->ellipsoidModel;

                            auto earth = vsg::TileDatabase::create();
                            earth->settings = settings;
                            earth->readDatabase(options);



                            


                            auto earth_scene = vsg::Group::create();
                            earth_scene->addChild(earth);

                            vsg::GeometryInfo geomInfo;


                            auto ecef = ellipsoidModel->convertLatLongAltitudeToECEF({25.686613, -100.316116, 0.0});
                            auto ecef_normal = vsg::normalize(ecef);
                            geomInfo.position = ecef;

                            geomInfo.dx.set(200000.0f, 0.0f, 0.0f);
                            geomInfo.dy.set(0.0f, 200000.0f, 0.0f);
                            geomInfo.dz.set(0.0f, 0.0f, 250000.0f);

                           auto box = builder->createCylinder(geomInfo, stateInfo);
                    


                           earth_scene->addChild(box);

**Reply by ztef** [reply]
the cylinder is oriented to the north,  i am trying to orient it parallell to the earth's surface

**Comment by robertosfield** [comment]
On Wed, 12 Jul 2023 at 22:28, esteban ortiz ***@***.***>
wrote:

> the cylinder is oriented to the north, i am trying to orient it parallell
> to the earth's surface
>
I would recommend creating the cylinder in a local coordinate frame, and
use a vsg::MatrixTransform as it's parent to place it in ECEF.  The
transform's matrix would be set using the
ElliposoidModel::computeLocalToWorldTransform(lat, long, altitude) value.

**Comment by ztef** [comment]
sounds good. I'll try that way.    Thank you in advance.

**Comment by ztef** [comment]
![Screenshot from 2023-07-12 16-19-35](https://github.com/vsg-dev/VulkanSceneGraph/assets/1700294/d710717a-a40a-41fc-8fb6-1f5800e80667)


It works !!!!    Thank you very much !!
